processing/src/main/java/io/druid/segment/StringDimensionIndexer.java

```diff
@@ -412,27 +415,27 @@ public IndexedInts getRow()
 
         int[] row = null;
         int rowSize = 0;
+
+        // usually due to currEntry's rowIndex is smaller than the row's rowIndex in which this dim first appears
         if (indices == null || indices.length == 0) {
           final int nullId = getEncodedValue(null, false);
           if (nullId > -1) {
-            if (nullId < maxId) {
-              row = new int[] {nullId};
-              rowSize = 1;
-            } else {
-              // Choose to use ArrayBasedIndexedInts later, instead of EmptyIndexedInts, for monomorphism
-              row = IntArrays.EMPTY_ARRAY;
-              rowSize = 0;
+            if (nullIdIntArray == null) {
+              nullIdIntArray = new int[] {nullId};
             }
+            row = nullIdIntArray;
```

Comment:
There is a race, the single element of `nullIdIntArray`, created in another thread, may not be visible in this thread yet. You can resolve this race by caching the whole `ArrayBasedIndexedInts` object, not just array.